### PR TITLE
support any kind of map in loops

### LINKED ISSUE #143: For loops over maps from a file seem to not work
When trying to do a for loop on values from a system property file, or from an input file to the CLI, I couldn't make it work.
When parsing the resource file, we get a *LinkedHashMap* object. 
We can use in the slang files some iterable over it like: entrySet(), keys() or values(), but all are non Serilaizable.
So I couldn't find a way to make it work...

### PR DESCRIPTION
**Pull request by Bonczidai**
fixes #143 

Signed-off-by: Bonczidai Levente <levente.bonczidai@hp.com>

### CODE REVIEW

**Comment by Bonczidai** [comment]
This approach simplifies the syntax from `k, v in map.items()` to `k, v in map` and works with any kind of object that implements the Map interface:

- when reading from file you get LinkedHashMap
- when using Python expression you get PyDictionary, both implementing Map

Internally we convert every Map to an Iterable object which is also Serializable -> an ArrayList of `org.apache.commons.lang3.tuple.Pair` elements.

@orius123 @oritstone let me know what you think..

**Comment by orius123** [comment]
sounds good to me.
why do you need to make it some python/java concrete map?
the only thing I don't like here is that it defers from python syntax..

**Comment by Bonczidai** [comment]
well, only Python 'maps' have `items()` method.. so if we want to support java maps we cannot stay with that syntax (maybe if we manually convert all java maps in the context to PyDictionary before looping task - I don't really like that approach)

**Comment by oritstone** [comment]
:+1:

**Comment by meirwah** [comment]
@Bonczidai please update branch...

**Comment by Bonczidai** [comment]
updated..

**Comment by meirwah** [comment]
Did you review this @orius123 @oritstone ?

**Comment by oritstone** [comment]
@Bonczidai is it ready for review?

**Comment by Bonczidai** [comment]
Yes

File: score-lang-compiler/src/test/java/org/openscore/lang/compiler/modeller/transformers/ForTransformerTest.java [hunk 1]
```diff
@@ -112,15 +112,6 @@ public void testValidMapStatementWithoutSpaceAfterComma() throws Exception {
     }
 
     @Test
-    public void testValidMapStatementWithExpression() throws Exception {
```
Comment (orius123) [inline score-lang-compiler/src/test/java/org/openscore/lang/compiler/modeller/transformers/ForTransformerTest.java]: why did you remove it?

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java [hunk 2]
```diff
@@ -16,15 +16,17 @@
 import org.openscore.lang.runtime.env.Context;
 import org.openscore.lang.runtime.env.ForLoopCondition;
 import org.openscore.lang.runtime.env.LoopCondition;
-import org.python.core.PyList;
 import org.python.core.PyObject;
-import org.python.core.PyTuple;
+import org.python.google.common.collect.Lists;
```
Comment (orius123) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: don't use jython shaded 3rd parties.
if you need guava add maven dep

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java [hunk 3]
```diff
@@ -86,10 +88,17 @@ private LoopCondition createForLoopCondition(ForLoopStatement forLoopStatement,
             throw new RuntimeException("Error evaluating for loop expression in task '" + nodeName + "', error is: \n" + t.getMessage(), t);
         }
 
-        if ((forLoopStatement instanceof MapForLoopStatement) && !(evalResult instanceof PyList)) {
-            throw new RuntimeException(
-                    "Invalid expression for iterating maps: " + collectionExpression
-                    + "\nValid format is: map.items()");
+        if (forLoopStatement instanceof MapForLoopStatement) {
+            if (evalResult instanceof Map) {
+                List<Map.Entry<Serializable, Serializable>> entriesAsSerializable = Lists.newArrayList();
```
Comment (orius123) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: String, Serializable?

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java [hunk 4]
```diff
@@ -86,10 +88,17 @@ private LoopCondition createForLoopCondition(ForLoopStatement forLoopStatement,
...
+                @SuppressWarnings("unchecked") Set<Map.Entry<Serializable, Serializable>> entrySet = ((Map) evalResult).entrySet();
+                for (Map.Entry<Serializable, Serializable> entry : entrySet) {
```
Comment (orius123) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: can't you construct a list from a set?

File: score-lang-compiler/src/test/java/org/openscore/lang/compiler/modeller/transformers/ForTransformerTest.java (hunk 1 above)
Comment (Bonczidai) [inline score-lang-compiler/src/test/java/org/openscore/lang/compiler/modeller/transformers/ForTransformerTest.java]: since we don't have `.items()` anymore, the tests simplifies to `k, v in dictionary` which is tested in  the testValidMapStatement() method

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java (hunk 3 above)
Comment (Bonczidai) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: Depends, do we want to admit non-String keys for maps in slang files: e.g. `k, v in { 1: "abc", 2: "efg" }` ? 

```
k=1
v="abc"
----------
k=2
v="egf"
```

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java (hunk 4 above)
Comment (Bonczidai) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: We need to make sure that `Pair` elements are created which are Serializable so we need to iterate the entrySet.. Maybe would be better to use lambda expression:

```
                @SuppressWarnings("unchecked") Set<Map.Entry<Serializable, Serializable>> entrySet = ((Map) evalResult).entrySet();

                List<Map.Entry<Serializable, Serializable>> entriesAsSerializable =
                        convert(entrySet, new Converter<Map.Entry<Serializable, Serializable>, Map.Entry<Serializable, Serializable>>() {
                    @Override
                    public Map.Entry<Serializable, Serializable> convert(Map.Entry<Serializable, Serializable> entry) {
                        return Pair.of(entry.getKey(), entry.getValue());
                    }
                });
```

WDYT?

Comment (Bonczidai) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: We have to stay with Pair elements, because `new ArrayList(entrySet)` returns an object that is not Serializable

**Comment by Bonczidai** [comment]
updated..

**Comment by meirwah** [comment]
can we merge this @orius123 ?

**Comment by orius123** [comment]
@Bonczidai delete the branch if you don't need it anymore.
